### ISSUE TITLE
runtime: rewrite gentraceback as an iterator API

### ISSUE THREAD

**Issue by aclements** [golang/go#54466]
Currently, all stack walking logic is in one venerable, large, and very, very complicated function: `runtime.gentraceback`. This function has three distinct operating modes: printing, populating a PC buffer, or invoking a callback. And it has three different modes of unwinding: physical Go frames, inlined Go frames, and cgo frames. It also has several flags. All of this logic is very interwoven.

I would like to replace all of this with a caller-driven iterator-style interface. This is a tracking issue for that change.

An iterator API will consolidate the logic for unwinding and allow us to lift out printing and pcbuf populating into separate code, while replacing the callback mode with direct use of the new API. It will allow us to better layer the different modes of unwinding by creating separate iterator types for physical, inlined, and cgo frames, while keeping the interface ergonomic. This is also a good opportunity to generally clean up this code.

As a follow-on, I plan to dramatically simplify the `defer` implementation. Regabi enabled many simplifications to `defer` and we've implemented many of them already, but there are more aggressive simplifications we haven't tackled yet. Part of this is simplifying open-coded defers, and doing that efficiently requires being able to simultaneously walk the stack frames and the defer stack. An iterator API will make this much easier to do.

An alternative approach would be to use a callback interface rather than an iterator. This would be an improvement over the status quo and also be a simpler change, but I think it has two drawbacks: 1. It makes layering physical/inlined frame unwinding more awkward because you need multiple levels of callbacks. 2. It's poorly suited to parallel iteration like we need for the open-coded defer implementation. Long term, an iterator API probably makes it simpler to [scan a goroutine stack while the goroutine is still running](https://github.com/golang/proposal/blob/master/design/17503-eliminate-rescan.md#low-pause-stack-scans) (reducing per-goroutine latency for goroutines with large stacks) because we can easily pause and resume unwinding, while a callback API doesn't easily afford this opportunity.

**Comment by randall77** [comment]
Keep in mind #7181, where we want to print bottom and top of deep stacks. It probably requires walking the stack twice, or keeping a buffer of 50 frames, or something like that.
I think it would not be a problem with this plan, just FYI.

**Comment by aclements** [comment]
> Keep in mind https://github.com/golang/go/issues/7181, where we want to print bottom and top of deep stacks.

I think this will actually make that dramatically easier by inverting the flow of traceback printing. With this change, the printer will drive the stack walk instead of the other way around, so I think it will be much easier for the printer to keep the buffer it needs, and to do so without adding complexity around the stack walk itself.

**Comment by aclements** [comment]
> As a follow-on, I plan to dramatically simplify the defer implementation.

Note to self: simplifying defer using an iterator-style gentraceback would also put us in a much better position to fix panic/recover causing bad stacks and memory leaks in the race detector (#26813, #37233).

**Comment by aclements** [comment]
Note to self: an iterator-style traceback would make it much easier to eliminate the limit on CPU profile stack frames (#56029).

**Comment by felixge** [comment]
I hacked on this a bit last week, see: https://github.com/felixge/go/pull/4

It's still far away from being mergable, and perhaps the WIP code is even worse than the initial code right now. I was trying to keep the current behavior exactly as is (bug-for-bug compatibility), but this lead to some gnarly hacks around write barriers that can probably be removed again in the end.

But I did manage to make quite a bit of progress towards inverting the loop control by taking many small steps while passing the test suite for each commit. So if anybody wants to take a look and leave some high-level comments (e.g. on whether or not a state machine inside the iterator is overkill or not), I'd be grateful.

I don't know when I'll find time to continue this work (or if I should), but the next step would be to finish moving the `pcbuf` filling into `gentraceback2` while also dealing with the remaining for loop that handles cgo traceback. After that it should be fairly easy to also extract the printing use case.

**Comment by aclements** [comment]
@felixge , I just mailed out [the work-in-progress CL](https://go.dev/cl/458218) I got to a few months ago before this got bumped by higher priorities. It's a pretty different approach from the one you took, though I ran into the same issues with write barriers, which I was in the process of engineering around when I put the work on hold.

**Comment by aclements** [comment]
I should add that the other thing I'd really like to do as part of this is to improve the test coverage of tracebacks. Right now we have decent coverage of the common path, but pretty poor coverage of the many, many corner cases.

**Comment by felixge** [comment]
> @felixge , I just mailed out [the work-in-progress CL](https://go.dev/cl/458218) I got to a few months ago before this got bumped by higher priorities. It's a pretty different approach from the one you took, though I ran into the same issues with write barriers, which I was in the process of engineering around when I put the work on hold.

Thanks. On first glance this looks much better/cleaner 😅. I'll try to take a more careful look later on.

> I should add that the other thing I'd really like to do as part of this is to improve the test coverage of tracebacks. Right now we have decent coverage of the common path, but pretty poor coverage of the many, many corner cases.

Agreed! I was wondering if would be feasible to have test cases that try to unwind functions at interesting PCs. In my mind this would work by placing a trap in some pc of the function (int3/`0xCC`?) and then doing an unwind in the trap handler. This could even be done for every pc of a small function.

But I'm not sure how portable this would be (needs `ptrace(2)`?) and if we could really cover enough tricky cases with it, or if more machinery would be needed.

If you have specific ideas around this, I could probably make time to work on test cases and leave the unwinder refactoring itself in your hands.

**Comment by aclements** [comment]
I mentioned when we spoke yesterday, but will note for the record here, that my best idea for how to test this is to use `perf_event_open` to inject breakpoints to get signals from which we can backtrace. That's Linux-specific, but at least it's architecture-independent, unlike `ptrace` which is both OS and architecture-specific for these sorts of things.

I found my notes on the space of things that would be good to test:

Traceback starting points:
- G is blocked in syscall. G is blocked in scheduler. G has not run yet.
- nil function call.
- Trap in ARM atomic helpers.

Stack contents:
- Jump through morestack. Jump through systemstack. Stack has a sigpanic on it.

Things to check:
- Correct PC in trap and sigpanic case, including trap at function entry.
- Wrappers get elided, except when the wrapper panicked. For both pcbuf and printing.

I'm sure this is not complete. Ideally we'd have some coverage information to make sure we're actually hitting the corner cases we're trying to hit. I think the new coverage support in 1.20 should be able to do this.

**Comment by prattmic** [comment]
To add to traceback starting points:
* Trap in VDSO
* Trap in setup (vdsoSP/PC) of VDSO calls

**Comment by cherrymui** [comment]
Trap in (various places in) function prologue is also an interesting place to test. Maybe also epilogue, although that is generally much safer/better.

Trap in injected functions (asyncPreempt, sigpanic) could also be interesting, although I think they have not caused much problem so far.

Maybe LR-machine specific: frameless leaf function could be one.

**Comment by aclements** [comment]
I've been digging more into testing tracebacks. I came up with a different approach that seems quite promising for testing the really dark corner cases: set your own single-step flag just before doing something we want to test, and in the SIGTRAP handler keep taking tracebacks and single-stepping until we get what we're looking for or return from the test situation. I have a rough prototype of this on x86 that works on Linux and may trivially work on other Unixes, and I suspect would work without much trouble on Windows. I'm pretty sure I can do it on ARM64, too, but haven't written the code. I'm not sure about other architectures.

However, I also looked at the coverage of the traceback code from the `runtime` and `runtime/pprof` tests and it's actually pretty good. Between that and @prattmic's assessment that the code change actually isn't too bad, I'm no longer worried that significantly beefing up the tests is a prerequisite. Better testing is always nice, of course.
